frontend/frontend_test.go

```diff
@@ -1,8 +1,476 @@
+	if err == nil {
+		testT.Fatal("GetTransaction unexpectedly succeeded")
+	}
+	if err.Error() != "Please call GetTransaction with txid" {
+		testT.Fatal("GetTransaction unexpected error message")
+	}
+	if rawtx != nil {
+		testT.Fatal("GetTransaction non-nil rawtx returned")
+	}
+
+	rawtx, err = lwd.GetTransaction(context.Background(),
+		&walletrpc.TxFilter{Block: &walletrpc.BlockID{Hash: []byte{}}})
+	if err == nil {
+		testT.Fatal("GetTransaction unexpectedly succeeded")
+	}
+	if err.Error() != "Can't GetTransaction with a blockhash+num. Please call GetTransaction with txid" {
+		testT.Fatal("GetTransaction unexpected error message")
+	}
+	if rawtx != nil {
+		testT.Fatal("GetTransaction non-nil rawtx returned")
+	}
+}
+
+func getblockStub(method string, params []json.RawMessage) (json.RawMessage, error) {
+	step++
+	var height string
+	err := json.Unmarshal(params[0], &height)
+	if err != nil {
+		testT.Fatal("could not unmarshal height")
+	}
+	if height != "1234" {
+		testT.Fatal("unexpected getblock height", height)
+	}
+
+	// Test retry logic (for the moment, it's very simple, just one retry).
+	switch step {
+	case 1:
+		return blocks[0], nil
+	case 2:
+		return nil, errors.New("getblock test error")
+	}
+	testT.Fatal("unexpected call to getblockStub")
+	return nil, nil
+}
+
+func TestGetLatestBlock(t *testing.T) {
+	testT = t
+	common.RawRequest = getblockStub
+	// This argument is not used (it may be in the future)
+	req := &walletrpc.ChainSpec{}
+
+	blockID, err := lwd.GetLatestBlock(context.Background(), req)
+	if err == nil {
+		t.Fatal("GetLatestBlock should have failed, empty cache")
+	}
+	if err.Error() != "Cache is empty. Server is probably not yet ready" {
+		t.Fatal("GetLatestBlock incorrect error", err)
+	}
+	if blockID != nil {
+		t.Fatal("unexpected blockID", blockID)
+	}
+
+	// This does zcashd rpc "getblock", calls getblockStub() above
+	block, err := common.GetBlock(cache, 1234)
+	if err != nil {
+		t.Fatal("getBlockFromRPC failed", err)
+	}
+	reorg, err := cache.Add(40, block)
+	if reorg {
+		t.Fatal("unexpected reorg")
+	}
+	if err != nil {
+		t.Fatal("cache.Add failed:", err)
+	}
+	blockID, err = lwd.GetLatestBlock(context.Background(), req)
+	if err != nil {
+		t.Fatal("lwd.GetLatestBlock failed", err)
+	}
+	if blockID.Height != 40 {
+		t.Fatal("unexpected blockID.height")
+	}
+	step = 0
+}
+
+// A valid address starts with "t", followed by 34 alpha characters;
+// these should all be detected as invalid.
+var addressTests = []string{
+	"",                                     // too short
+	"a",                                    // too short
+	"t123456789012345678901234567890123",   // one byte too short
+	"t12345678901234567890123456789012345", // one byte too long
+	"t123456789012345678901234567890123*",  // invalid "*"
+	"s1234567890123456789012345678901234",  // doesn't start with "t"
+	" t1234567890123456789012345678901234", // extra stuff before
+	"t1234567890123456789012345678901234 ", // extra stuff after
```

Comment:
This test case addresses https://github.com/adityapk00/lightwalletd/pull/4

Reply:
A test case for the newline thing would be "\nt1234567890123456789012345678901234".